azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/KeyVaultOperation.java

```diff
@@ -3,91 +3,148 @@
-import org.springframework.util.StringUtils;
 
-import java.util.ArrayList;
-import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
-import java.util.Optional;
-import java.util.stream.Collectors;
+import java.util.Timer;
+import java.util.TimerTask;
 import java.util.stream.Stream;
 
 @Slf4j
 public class KeyVaultOperation {
-    
+
     /**
      * Stores the case sensitive flag.
      */
     private final boolean caseSensitive;
 
-    private final SecretClient keyVaultClient;
-    private final String vaultUri;
-    private volatile List<String> secretNames;
-    private final boolean secretNamesAlreadyConfigured;
-    private final long secretNamesRefreshIntervalInMs;
-    private volatile long secretNamesLastUpdateTime;
+    /**
+     * Stores the properties.
+     */
+    private LinkedHashMap<String, String> properties = new LinkedHashMap<>();
+
+    /**
+     * Stores the secret client.
+     */
+    private final SecretClient secretClient;
+
+    /**
+     * Stores the secret keys.
+     */
+    private List<String> secretKeys;
 
+    /**
+     * Constructor.
+     *
+     * @param secretClient the Key Vault secret client.
+     * @param refreshInMillis the refresh in milliseconds (0 or less disables
+     * refresh).
+     * @param caseSensitive the case sensitive flag.
+     */
     public KeyVaultOperation(
-            final SecretClient keyVaultClient,
-            String vaultUri,
-            final long secretKeysRefreshIntervalInMs,
-            final List<String> secretNames,
-            boolean caseSensitive
-    ) {
+            final SecretClient secretClient,
+            final long refreshInMillis,
+            List<String> secretKeys,
+            boolean caseSensitive) {
+
         this.caseSensitive = caseSensitive;
-        this.keyVaultClient = keyVaultClient;
-        // TODO(pan): need to validate why last '/' need to be truncated.
-        this.vaultUri = StringUtils.trimTrailingCharacter(vaultUri.trim(), '/');
-        this.secretNames = Optional.ofNullable(secretNames)
-                .map(Collection::stream)
-                .orElseGet(Stream::empty)
-                .map(this::toKeyVaultSecretName)
-                .distinct()
-                .collect(Collectors.toList());
-        this.secretNamesAlreadyConfigured = !this.secretNames.isEmpty();
-        this.secretNamesRefreshIntervalInMs = secretKeysRefreshIntervalInMs;
-        this.secretNamesLastUpdateTime = 0;
+        this.secretClient = secretClient;
+        this.secretKeys = secretKeys;
+
+        if (refreshInMillis > 0) {
+            final Timer timer = new Timer();
+            final TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    refreshProperties();
+                }
+            };
+            timer.scheduleAtFixedRate(task, 0, refreshInMillis);
+        } else {
+            refreshProperties();
```

Comment:
If `refreshInMillis <= 0`, we will not use cache, we should get the result(secret name list and secret value) from key-vault in every request.

Reply:
> Should we allow for no caching at all?

I already discussed with @jialindai , we think that we allow no caching at all. 

Let user configure whether need to cache.

If user configure `refreshInMillis == 0`, not cache make sense. 

Maybe can can use `kay-vault.cache.ms` instead of `refresh-interval` as configure item?

Reply:
No caching will be detrimental to any application as you are not in control of when the PropertySource gets consulted. I strongly advice against that.

Reply:
Your concern make sense.

But current behavior is no cache. 

IMU,
* About 80%(just guess) of key-vault property will only get one time. (when start the app).
* It's OK to let our customer determine whether to use cache.

Reply:
If we set the refresh interval to -1 to indicate we do NO caching at all then I will go ahead and implement that, but I still think this is bad from a user experience point of view. Our code now can slow down any part of their application because we do NOT control the order in which PropertySources are consulted, nor at what time! So agreed on the -1?

Reply:
As discussed in email. Let our customer to decide the TTL of cache.